rpcs3/Emu/Cell/Modules/cellGem.cpp

```diff
@@ -264,6 +264,37 @@ struct gem_config_data
 
 	u64 start_timestamp_us = 0;
 
+	atomic_t<u32> m_wake_up = 0;
+	atomic_t<u32> m_done = 0;
+
+	void wake_up()
+	{
+		m_wake_up.release(1);
+		m_wake_up.notify_one();
+	}
+
+	void done()
+	{
+		m_done.release(1);
+		m_done.notify_one();
+	}
+
+	void wait_for_result()
+	{
+		if (!m_done)
+		{
+			m_done.wait(0);
+			m_done = 0;
+		}
+	}
+
+	gem_config_data& operator=(thread_state)
+	{
+		wake_up();
+		done();
```

Comment:
shouldn't done() be called exclusive by the thread? In case of emulation stop, PPUs should check their own state and not depend on notifications from others. 

1. Use thread_ctrl::wait_on() instead of atomic_t::wait
2. Check ppu_thread::state if stopped before wait.
3. Remove this line.
4. If PPU wakes up prematurely on emulation stopping, add cpu_state::again to state.

Reply:
I'm assuming the = operators aren't necessary anymore then